### PR TITLE
Implement scipy.sparse.linalg.cg (second try)

### LINKED ISSUE #2557: Inverse Fisher/Hessian as GD Preconditioner
Hi all,

Following up on #2197, I was hoping to implement natural gradient descent or Gauss-Newton for DNNs, for which I have to invert a large p-by-p matrix, where p is the number of parameters of my neural network. I understand that if I only want to perform a small number of gradient steps, I could implement H^-1 * g by using conjugate gradient descent, but I'm not sure how to best do that in JAX. Is there an existing implementation of CG in JAX? I haven't been able to find one. In general, am I right to assume that using the full inverse Hessian as a preconditioner should be feasible with CG in JAX?

Thank you!

### PR DESCRIPTION
**Pull request by shoyer**
Works exactly like SciPy's `cg` except you need need to supply linear operators as functions instead of `LinearOperator`. I didn't feel like it was worth the trouble to make a `LinearOperator` abstraction for JAX yet, but that would be easy to add later.

Here's the function signature (adapted from SciPy with minor tweaks to ensure correctness for JAX):
```python
def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
  """Use Conjugate Gradient iteration to solve ``Ax = b``.

  The numerics of JAX's ``cg`` should exact match SciPy's ``cg`` (up to
  numerical precision), but note that the interface is slightly different: you
  need to supply the linear operator ``A`` as a function instead of sparse
  matrix or ``LinearOperator``.

  Parameters
  ----------
  A : function
      Function that calculate the matrix-vector product ``Ax`` when called like
      ``A(x)``. ``A`` must represent a hermitian, positive definite matrix.
  b : array
      Right hand side of the linear system. Has shape (N,).

  Returns
  -------
  x : array
      The converged solution.
  info : None
      Placeholder for convergence information. In the future, JAX will report
      the number of iterations when convergence is not achieved, like SciPy.

  Other Parameters
  ----------------
  x0 : array
      Starting guess for the solution.
  tol, atol : float, optional
      Tolerances for convergence, ``norm(residual) <= max(tol*norm(b), atol)``.
      We do not implement SciPy's "legacy" behavior, so JAX's tolerance will
      differ from SciPy unless you explicitly pass ``atol`` to SciPy's ``cg``.
  maxiter : integer
      Maximum number of iterations.  Iteration will stop after maxiter
      steps even if the specified tolerance has not been achieved.
  M : function
      Preconditioner for A.  The preconditioner should approximate the
      inverse of A.  Effective preconditioning dramatically improves the
      rate of convergence, which implies that fewer iterations are needed
      to reach a given error tolerance.

  See also
  --------
  scipy.sparse.linalg.cg
  """
```

Based on @TuanNguyen27's work in https://github.com/google/jax/pull/1821 (thanks for getting this started!)

Fixes https://github.com/google/jax/issues/2557

xref https://github.com/google/jax/issues/1032, https://github.com/google/jax/issues/1531

### CODE REVIEW

**Comment by shoyer** [comment]
@TuanNguyen27 could you OK using your commits in this PR? See the note above from @googlebot for instructions.

**Comment by TuanNguyen27** [comment]
@googlebot I consent.

**Review by mattjj — APPROVED** [review]
Beautiful code, as always. This is going to be super valuable for a lot of users right away.
